File: momenttensor/matlab/nualpha2lam.m
function lam = nualpha2lam(nu,alpha)
%NUALPHA2LAM converts nu and alpha to unit lambda vector
%
% INPUT
%   nu      n-dimensional vector of Poisson ratios (unitless)
%   alpha   n-dimensional vector of angles between fault normal and slip vector, degrees [0,180]
%
%  OUTPUT
%   lam     3 x n set of normalized eigenvalue triples, sorted lam1 >= lam2 >= lam3
%
% Reverse function is lam2nualpha.m
% See Tape and Tape (2013), "The classical model for moment tensors"
% 
% Carl Tape, 08-Jan-2013
%

% row vectors
alpha = alpha(:)';
nu = nu(:)';

cosa = cos(alpha*pi/180);

% TapeTape2013, Eq 30
lam1 = cosa ./ (1-2*nu) + 1;
lam2 = 2 .* nu .* cosa ./ (1-2*nu);
lam3 = cosa ./ (1-2*nu) - 1;

% unit norm
lam = [lam1 ; lam2 ; lam3];
mag = sqrt( lam1.^2 + lam2.^2 + lam3.^2 );
lam = lam ./ repmat(mag,3,1);
%mag = sqrt( lam(1,:).^2 + lam(2,:).^2 + lam(3,:).^2 );  % check

%==========================================================================
% EXAMPLE

if 0==1
    %% grid of nu-alpha values
    avec = [5:10:175];           % avoid alpha = 90 (nu nonunique)
    nvec = [-1:0.05:0.45];       % restrict nu to permissible range
    [nu,alpha] = meshgrid(nvec,avec);
    nu = nu(:); alpha = alpha(:);
    lam = nualpha2lam(nu,alpha);
    
    % check
    [nucheck,alphacheck] = lam2nualpha(lam);
    norm(nucheck - nu)
    norm(alphacheck - alpha)
    
    %% CDC arc for nu=0.25
    avec = [0:2:180];
    lam = nualpha2lam(0.25,avec);
    [gamma,delta,M0,mu,lamdev,lamiso] = lam2lune(lam);
    figure; plot(gamma,delta,'.'); axis([-30 30 -90 90]);
end

%==========================================================================

File: momenttensor/c_src/nualpha2lam.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "compearth.h"

/*!
 * @brief Converts \f$ \nu \f$ and \f$ \alpha \f$ to unit \f$ \Lambda \f$
 *        vector.  This is hte inverse function of lam2nualpha.  See
 *        Tape and Tape (2013), "The classical model for moment tensors".
 *
 * @param[in] n       Number of eigenvalue triples.
 * @param[in] nu      Unitless Poisson parameters.  This is an array of
 *                    dimension n.
 * @param[in] alpha   Angles between fault normal and slip vector
 *                    where \f$ \alpha \in [0,180] \f$.  This is an array
 *                    of dimension n.
 *
 * @param[out] lam    [3 x n] set of eigenvalue triples.  The leading dimension
 *                    is 3.
 *
 * @author Carl Tape and converted to C by Ben Baker
 *
 * @copyright MIT
 *
 */
int compearth_nualpha2lam(const int n,
                          const double *__restrict__ nu,
                          const double *__restrict__ alpha,
                          double *__restrict__ lam)
{
    double cos_1p2n, cosa, lam1, lam2, lam3, mag;
    int i;
    const double rad = M_PI/180.0;
    if (n < 1 || lam == NULL || alpha == NULL || nu == NULL)
    {   
        if (n < 1){fprintf(stderr, "%s: Error no moment tensors\n", __func__);}
        if (lam == NULL){fprintf(stderr, "%s: Error lam is NULL\n", __func__);}
        if (alpha == NULL)
        {
            fprintf(stderr, "%s: Error alpha is NULL\n", __func__);
        }
        if (nu == NULL){fprintf(stderr, "%s: Error nu is NULL\n", __func__);} 
        return -1;
    }
    for (i=0; i<n; i++)
    {
        cosa = cos(alpha[i]*rad);
        // TapeTape2013 Eq 30
        cos_1p2n = cosa/(1.0 - 2.0*nu[i]);
        lam1 = cos_1p2n + 1.0;
        lam2 = 2.*nu[i]*cos_1p2n;
        lam3 = cos_1p2n - 1.0;
        // Unit norm
        mag = sqrt(lam1*lam1 + lam2*lam2 + lam3*lam3);
        lam[3*i+0] = lam1/mag;
        lam[3*i+1] = lam2/mag;
        lam[3*i+2] = lam3/mag; 
    }
    return 0;
}
